### PR TITLE
Support subgraph datasources

### PR DESCRIPTION
**Pull request by incrypto32**
This PR adds new types for subgraph datasources to graph-ts and codegen functionality to generate schema code for the source subgraphs.
Detailed docs about subgraph datasources [here](https://github.com/incrypto32/subgraph-composition-sample-subgraph/tree/main/docs).

This PR 

1.  Adds support to parse the new type of datasource in the manifest
2. Adds the new `EntityTrigger` type to `graph-ts` which is the Assemblyscript represenation of entity triggers.
3. Adds an enum `EntityOp` which represents the entity operations done on the source subgraph
4. Runs codegen for the schema of source subgraphs so that dependant subgraph also has access to entity types from source subgraphs

### CODE REVIEW

**Comment by 0237h** [comment]
Hey @incrypto32, thanks for the PR !

Gonna take a bit of time to review. In the meantime if you can provide some more context/information as to what these feature would bring to developers, would be very much appreciated thanks :+1:

**Comment by 0237h** [comment]
Given the context, we'll also wait for some of the [related PRs](https://github.com/search?q=repo%3Agraphprotocol%2Fgraph-node+composition++&type=pullrequests&state=open) in [`graph-node`](https://github.com/graphprotocol/graph-node/) to be merged before merging this one as well.

**Comment by incrypto32** [comment]
Hey @0237h , i'll write up the required context soon. TL;DR is this will support the code generation and the new type of datasource thats being added as part of the subgraph composition feature. As you mentioned. Yes those graph-node PR's need to be merge before getting this out. I'll write a detailed comment on the changes later.

**Comment by incrypto32** [comment]
@0237h I just added a description to the PR. let me know if more info is needed

File: packages/cli/src/type-generator.ts [hunk 1]
```diff
@@ -106,6 +144,37 @@ export default class TypeGenerator {
     }
   }
 
+  async loadSubgraphSchemaFromIPFS(ipfsClient: any, manifest: string) {
```
Comment (0237h) [inline packages/cli/src/type-generator.ts:147]: Could we add some tests for this method ?

File: packages/cli/src/protocols/index.ts [hunk 2]
```diff
@@ -290,6 +299,21 @@ const ethereumProtocol: ProtocolConfig = {
...
+  contract: undefined,
+  getTemplateCodeGen: undefined,
+  getTypeGenerator(options) {
+    return new EthereumTypeGenerator(options);
+  },
+  getSubgraph(options) {
+    return new SubgraphDS(options);
+  },
+  manifestScaffold: undefined,
+  mappingScaffold: undefined,
```
Comment (0237h) [inline packages/cli/src/protocols/index.ts:314]: Any reason we're not using the [subgraph scaffolds](https://github.com/graphprotocol/graph-tooling/tree/b45cf7361075df227c4a3699d884070c02d121b5/packages/cli/src/protocols/subgraph/scaffold) here ?

File: packages/cli/src/protocols/subgraph/subgraph.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import immutable from 'immutable';
+import { Subgraph, SubgraphOptions } from '../subgraph';
+
+export default class SubgraphDS implements Subgraph {
```
Comment (0237h) [inline packages/cli/src/protocols/subgraph/subgraph.ts:4]: I think `SubgraphDataSource` would be more explicit.

**Review by 0237h — CHANGES_REQUESTED** [review]
@incrypto32 Thanks for the updated description !

A few minor comments, other LGTM.
Do you know if the `graph-node` integration for subgraph datasources is also ready ?

**Comment by YaroShkvorets** [comment]
Doesn't seem to work for me. After running `graph init` and selecting "subgraph" as a protocol it tries to download ABI for I don't know what. I assume it shouldn't, since I never specified the contract address?
What's the flow here @incrypto32 ? 

image

**Comment by incrypto32** [comment]
Oops, i think i didn't check the `graph init` support just the codegen and build. I tested with an existing subgraph that i modified to use the new datasource. 
Thanks for testing it out @YaroShkvorets, i'll add the init support too.

**Comment by incrypto32** [comment]
@YaroShkvorets just added the init command. Please re review

**Comment by YaroShkvorets** [comment]
```
❯ ./bin/run init --skip-install --skip-git
 ›   Warning: @graphprotocol/graph-cli update available from 0.87.0 to 0.90.1.
 ›   Warning: In next major version, this flag will be removed. By default we will stop initializing a Git repository.
✔ Protocol · subgraph
✔ Subgraph slug · asdf
✔ Directory to create the subgraph in · asdf
? Subgraph network …
? Subgraph network …
? Subgraph network …
✔ Subgraph network · mainnet
✔ Source subgraph identifier · QmTZ8ejXJxRo7vDBS4uwqBeGoxLSWbhaA7oXa1RvxunLy7
✔ IPFS node to use for fetching subgraph manifest · https://api.thegraph.com/ipfs/api/v0
✔ Start Block · 0
  Generate subgraph
✖ Failed to create subgraph scaffold: Cannot read properties of undefined (reading 'length')
    TypeError: Cannot read properties of undefined (reading 'length')
```

**Comment by YaroShkvorets** [comment]
Let's merge it @0237h , then let @incrypto32 polish out any bugs.
There is a lot of overlap with what I'm doing so I don't want to keep this PR open much longer to avoid merge hell.

**Review by YaroShkvorets — APPROVED** [review]

**Comment by incrypto32** [comment]
@YaroShkvorets as will this be released soon?
Actually we need to wait before releasing this for the graph-node PR's to be merged. cc: @alex-pakalniskis 
https://github.com/graphprotocol/graph-tooling/pull/1754#issuecomment-2462599489

**Comment by YaroShkvorets** [comment]
Yes, let's coordinate on the release.

**Pull request merged**